File: app/src/main/java/app/quranhub/data/repository/RecitationsRepository.java
package app.quranhub.data.repository;

import androidx.annotation.NonNull;

import com.google.firebase.firestore.FirebaseFirestore;

import java.util.List;

import app.quranhub.data.model.ReciterModel;
import io.reactivex.Single;

public class RecitationsRepository {

    private static final String TAG = RecitationsRepository.class.getSimpleName();

    private final FirebaseFirestore db = FirebaseFirestore.getInstance();

    @NonNull
    public Single<List<ReciterModel>> getRecitersForRecitation(@NonNull String recitationKey) {
        return Single.create(emitter -> {
            db.collection("recitations")
                    .document(recitationKey)
                    .collection("reciters")
                    .get()
                    .addOnCompleteListener(task -> {
                        if (task.isSuccessful()) {
                            final List<ReciterModel> reciterModels = task.getResult().toObjects(ReciterModel.class);
                            emitter.onSuccess(reciterModels);
                        } else {
                            emitter.onError(task.getException());
                        }
                    });
        });
    }

}

File: app/src/main/java/app/quranhub/data/repository/RecitationsRepository.kt
package app.quranhub.data.repository

import app.quranhub.data.model.ReciterModel
import com.google.android.gms.tasks.Task
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.QuerySnapshot
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class RecitationsRepository {

    private val db = FirebaseFirestore.getInstance()

    suspend fun getRecitersForRecitation(recitationKey: String): List<ReciterModel> {
        return suspendCancellableCoroutine { continuation ->
            db.collection("recitations")
                .document(recitationKey)
                .collection("reciters")
                .get()
                .addOnCompleteListener { task: Task<QuerySnapshot> ->
                    if (task.isSuccessful) {
                        continuation.resume(
                            task.result.toObjects(ReciterModel::class.java)
                        )
                    } else {
                        continuation.resumeWithException(
                            task.exception
                                ?: IllegalStateException("Firestore task failed without exception")
                        )
                    }
                }
        }
    }
}
